src/services/refactors/moveToNewFile.ts

```diff
@@ -0,0 +1,621 @@
+
+        const currentDirectory = getDirectoryPath(oldFile.fileName);
+        const extension = extensionFromPath(oldFile.fileName);
+        const newModuleName = makeUniqueModuleName(getNewModuleName(usage.movedSymbols), extension, currentDirectory, host);
+        const newFileNameWithExtension = newModuleName + extension;
+
+        // If previous file was global, this is easy.
+        changes.createNewFile(oldFile, combinePaths(currentDirectory, newFileNameWithExtension), getNewStatements(oldFile, usage, changes, toMove, program, newModuleName));
+
+        addNewFileToTsconfig(program, changes, normalizePath(combinePaths(oldFile.fileName, "..", newFileNameWithExtension)));
+    }
+
+    function addNewFileToTsconfig(program: Program, changes: textChanges.ChangeTracker, newFilePath: string): void {
+        const cfg = program.getCompilerOptions().configFile;
+        const filesProp = cfg && cfg.jsonObject && find(cfg.jsonObject.properties, (prop): prop is PropertyAssignment =>
+            isPropertyAssignment(prop) && isStringLiteral(prop.name) && prop.name.text === "files");
+        if (filesProp && isArrayLiteralExpression(filesProp.initializer)) {
+            changes.insertNodeInListAfter(cfg, last(filesProp.initializer.elements), createLiteral(newFilePath), filesProp.initializer.elements);
+        }
+    }
+
+    function getNewStatements(
+        oldFile: SourceFile, usage: UsageInfo, changes: textChanges.ChangeTracker, toMove: ReadonlyArray<Statement>, program: Program, newModuleName: string,
+    ): ReadonlyArray<Statement> {
+        const checker = program.getTypeChecker();
+
+        if (!oldFile.externalModuleIndicator && !oldFile.commonJsModuleIndicator) {
+            changes.deleteNodeRange(oldFile, first(toMove), last(toMove));
+            return toMove;
+        }
+
+        const useEs6ModuleSyntax = !!oldFile.externalModuleIndicator;
+        const importsFromNewFile = createOldFileImportsFromNewFile(usage.oldFileImportsFromNewFile, newModuleName, useEs6ModuleSyntax);
+        if (importsFromNewFile) {
+            changes.insertNodeBefore(oldFile, oldFile.statements[0], importsFromNewFile, /*blankLineBetween*/ true);
+        }
+
+        deleteUnusedOldImports(oldFile, toMove, changes, usage.unusedImportsFromOldFile, checker);
+        changes.deleteNodeRange(oldFile, first(toMove), last(toMove));
+
+        updateImportsInOtherFiles(changes, program, oldFile, usage.movedSymbols, newModuleName);
+
+        return [
+            ...getNewFileImportsAndAddExportInOldFile(oldFile, usage.oldImportsNeededByNewFile, usage.newFileImportsFromOldFile, changes, checker, useEs6ModuleSyntax),
+            ...addExports(oldFile, toMove, usage.oldFileImportsFromNewFile, useEs6ModuleSyntax),
+        ];
+    }
+
+    function deleteUnusedOldImports(oldFile: SourceFile, toMove: ReadonlyArray<Statement>, changes: textChanges.ChangeTracker, toDelete: ReadonlySymbolSet, checker: TypeChecker) {

```

Comment:
I think the goal of this function is to remove the imports that were only needed for the code being moved.  It seems though, like it might also remove imports that were unused to begin with.  Personally, I think it seems strange to do a partial Organize Imports as part of this operation.  If we are going to do so, then my preference would be to have the editor trigger (unless the server has some extra knowledge?) so that it can appear separately on the undo stack.

Reply:
`unusedImportsFromOldFile` will be a subset of `oldImportsNeededByNewFile`, so we won't remove purely-unused imports.